### PR TITLE
Update cross-shard deposit schema

### PR DESCRIPTION
**Pull request by ninjaahhh**
tweaked serialization/deserialization of xshard tx deposit because of schema change

### CODE REVIEW

File: quarkchain/evm/messages.py [hunk 1]
```diff
@@ -407,102 +406,95 @@ def apply_transaction(state, tx: transactions.Transaction, tx_wrapper_hash):
...
-                state.qkc_config.ENABLE_EVM_TIMESTAMP is None
-                or state.timestamp >= state.qkc_config.ENABLE_EVM_TIMESTAMP
-            ):
-                remote_gas_reserved = tx.startgas - intrinsic_gas
-            ext.add_cross_shard_transaction_deposit(
-                quarkchain.core.CrossShardTransactionDeposit(
-                    tx_hash=tx_wrapper_hash,
-                    from_address=quarkchain.core.Address(
-                        tx.sender, tx.from_full_shard_key
+    if not tx.is_cross_shard:
```
Comment (ninjaahhh) [inline quarkchain/evm/messages.py]: changed the order of control flow for better readability. no actual change to the logic

File: quarkchain/core.py [hunk 2]
```diff
@@ -1102,7 +1102,7 @@ def __init__(
         self.is_from_root_chain = is_from_root_chain
 
 
-class CrossShardTransactionDeposit(Serializable):
+class CrossShardTransactionDepositV0(Serializable):
```
Comment (qizhou) [inline quarkchain/core.py]: For versioning, I would suggest
- CrossShardTransactionDepositDeprecated as V0
- CrossShardTransactionDeposit as V1
- CrossShardTransactionDeposit with the refund field as V2

File: quarkchain/core.py [hunk 3]
```diff
@@ -1162,49 +1171,76 @@ def __init__(self, tx_list):
...
-                    create_contract=False,
-                    is_from_root_chain=tx.is_from_root_chain,
-                    refund_rate=100,
-                )
-                for tx in old_list.tx_list
-            ]
-        )
+    @classmethod
+    def from_data(cls, data):
+        size = ByteBuffer(data).get_uint(4)
```
Comment (qizhou) [inline quarkchain/core.py]: This will throw if data is empty?

File: quarkchain/core.py [hunk 4]
```diff
@@ -1162,49 +1171,76 @@ def __init__(self, tx_list):
...
+                        is_from_root_chain=tx.is_from_root_chain,
+                        refund_rate=100,
+                    )
+                    for tx in old_list.tx_list
+                ]
+            )
+
+        if version == 0:
+            v0_list = CrossShardTransactionListV0.deserialize(data)
+            # magic!
```
Comment (qizhou) [inline quarkchain/core.py]: I think we should use the similar logic for DeprecatedList for consistency.

File: quarkchain/core.py (hunk 3 above)
Comment (ninjaahhh) [reply quarkchain/core.py]: it's the same logic as previous `from_data`: `is_old_list` will also read `size` first. so I think it probably will not throw

File: quarkchain/core.py (hunk 4 above)
Comment (ninjaahhh) [reply quarkchain/core.py]: since we only added one field for V2, I think this is more readable and concise

File: quarkchain/core.py (hunk 2 above)
Comment (ninjaahhh) [reply quarkchain/core.py]: good call. will do

Comment (ninjaahhh) [reply quarkchain/core.py]: wait, it doesn't make sense, as in line 1182 we are currently writing version 0 to all the deposits

File: quarkchain/core.py (hunk 4 above)
Comment (ninjaahhh) [reply quarkchain/core.py]: updated

**Comment by ninjaahhh** [comment]
updated, take another look when you have time @qizhou

**Pull request merged**